### ISSUE TITLE
Using Mockery with Lumen framework (Laravel)

### ISSUE THREAD

**Issue by KeyWeeUsr** [mockery/mockery#984]
I haven't found any chat or something, so I'll post it here.

Is there anyone who successfully mocked Eloquent models in Laravel or Lumen? I have tried multiple mocking libraries and no matter what I try, it still doesn't work and happily proceeds to the original functions. As in [this question](https://stackoverflow.com/q/49641751/5994041) direct mocking does not work, aliasing does not work, basically nothing works at all.

Any ideas? Could someone please give an example for mocking the basic Lumen project (`composer global require "laravel/lumen-installer"`) + Eloquent model? Or Laravel + Eloquent?

**Comment by mfn** [comment]
I don't think you realistically can (Laravel user here).

From the example on SO:
```php
        $user = UserModel::find($id); // mock here**************************************
…
        $res = UserModel::where('id', $id)
               ->update(['password' => $password]); // mock here*****************************
```

You can't mock these static calls, they are not even defined methods by that name, they trigger `\Illuminate\Database\Eloquent\Model::__callStatic` which immediately goes to `\Illuminate\Database\Eloquent\Model::__call` which forwards the call to a _newly_ created `\Illuminate\Database\Eloquent\Builder::__construct` (`find` is defined on `\Illuminate\Database\Eloquent\Builder::find` at the end of the day).

The aliasing workaround (probably) doesn't work because the model was already loaded for some other reason.

If you are really about to mock the DB away in this scenario, 
I suggest to write the code differently: use _some kind_ of abstraction in between.
Oh, and don't use static methods.

A simple no-state-singleton-repository class having all your queries defined which you inject in your `UserLogic` and then you kind mock what you want:

```php
// Name it what you want, it's not a "repository pattern" per se
class UserRepository {
  public function findById(int $id): User {
    …
  }
  public function updatePassword(int $id, string $password): User {
  }
}
```
```php
class UserLogic {
  private $userRepository;
  public function __construct(UserRepository $userRepository) {
    $this->userRepository = $userRepository;
  }
  public function updatePassword(int $id, string $password) {
  }
}
```

**Comment by KeyWeeUsr** [comment]
Thank you for the explanation it made me thinking and I figured it out. First, about the loaded class, I was using Lumen tutorial for TestCase:

```php
class TestCase extends \Laravel\Lumen\Testing\TestCase {
    function createApplication() {
        return require __DIR__ . '/../bootstrap/app.php';
    }
```

which most likely does the loading. I manually changed it to `PHPUnit\Framework\TestCase` for the test with mocks and the aliases started to work. For example:

```php
// PHPUnit\Framework\TestCase
$mock = Mockery::mock("alias:App\MyModel");  // autoloaded from "app" folder via composer PSR4
```

Re Eloquent static calls, damn, that's really nasty. I went through Lumen + Laravel + that Illuminate thingy (whatever that is in the relationship between those two) and I found multiple such cases in the source code. Lovely to debug... But that repository-like class is a very nice approach if you want to avoid doing static calls, +1. 

Still, avoiding static calls is IMO silly. I'm not going to wrap each static call from external libraries just because of that. It's the same nonsense as Java folks with "removing" pointers from the visible language and in the end still having errors with dereferencing them in the background and pointer-related exception names = just confusing people.

Instead of avoiding, I think the community should focus on making work with them easier (i.e. mocking in this case). Mockery does a great job! :clap:
